flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunnerTest.java

```diff
@@ -45,44 +43,46 @@
 	@Rule
 	public final Timeout timeout = Timeout.seconds(30);
 
+	private SystemExitTrackingSecurityManager systemExitTrackingSecurityManager;
+	private TaskManagerRunner taskManagerRunner;
+
+	@Before
+	public void before() {
+		systemExitTrackingSecurityManager = new SystemExitTrackingSecurityManager();
+		System.setSecurityManager(systemExitTrackingSecurityManager);
+	}
+
+	@After
+	public void after() throws Exception {
+		System.setSecurityManager(null);
+		if (taskManagerRunner != null) {
+			taskManagerRunner.close();
+		}
+	}
+
 	@Test
 	public void testShouldShutdownOnFatalError() throws Exception {
-		try (TaskManagerRunner taskManagerRunner = createTaskManagerRunner(createConfiguration())) {
-			taskManagerRunner.start();
+		taskManagerRunner = createTaskManagerRunner(createConfiguration());
 
-			final SystemExitTrackingSecurityManager systemExitTrackingSecurityManager =
-				runWithSystemExitTracking(() -> {
-					taskManagerRunner.onFatalError(new RuntimeException());
-					taskManagerRunner.getTerminationFuture().get(30, TimeUnit.SECONDS);
-				});
+		taskManagerRunner.onFatalError(new RuntimeException());
 
+		eventually(() -> {
```

Comment:
Instead of busy looping, I would prefer to change the `SystemExitTrackingSecurityManager` to return a `CompletableFuture<Integer>` which is completed with the exit code with which `System.exit(exitCode)` is called.